From 6641211323f0e1b705108c250a24349f7c2dd412 Mon Sep 17 00:00:00 2001
From: Bogdan Hamciuc <bogdan.hamciuc@nxp.com>
Date: Fri, 6 Dec 2019 14:17:56 +0200
Subject: [PATCH 036/269] s32g: Serial console setup

Enable serial output from LinFlexD_0. For now, only the standard output
is linked; the error console is dummy (will be implemented at a later
point), and serial input is not in our scope.
Also, UART parameters are hard-coded to match the hardware values that
other drivers (such as U-Boot's LinFlex UART driver) calculate
dynamically.

Upstream-Status: Pending 

Signed-off-by: Bogdan Hamciuc <bogdan.hamciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
Issue: ALB-4275
Issue: ALB-4276
---
 plat/s32g/include/s32g_linflexuart.h | 21 +++++++++------------
 plat/s32g/s32g_linflexuart.S         |  8 ++++----
 2 files changed, 13 insertions(+), 16 deletions(-)

diff --git a/plat/s32g/include/s32g_linflexuart.h b/plat/s32g/include/s32g_linflexuart.h
index 673b6dd90..2b4fb031d 100644
--- a/plat/s32g/include/s32g_linflexuart.h
+++ b/plat/s32g/include/s32g_linflexuart.h
@@ -12,19 +12,16 @@
 #define S32G_UART_BASE		S32G_LINFLEX0_BASE
 #define S32G_UART_SIZE		S32G_LINFLEX0_SIZE
 #define S32G_UART_BAUDRATE	(115200)
-/* TODO revisit this; for now we'll hard-code the clock/divider settings instead
- * of deriving them
- */
-#define S32G_UART_CLOCK_HZ		(133333333)
+#define S32G_UART_CLOCK_HZ	(133333333)
 
-#define S32G_LINFLEX_LINCR1		(0x0)
-#define S32G_LINFLEX_LINSR		(0x8)
-#define S32G_LINFLEX_UARTCR		(0x10)
-#define S32G_LINFLEX_UARTSR		(0x14)
-#define S32G_LINFLEX_LINIBRR		(0x28)
-#define S32G_LINFLEX_LINFBRR		(0x24)
-#define S32G_LINFLEX_BDRL		(0x38)
-#define S32G_LINFLEX_UARTPTO		(0x50)
+#define S32G_LINFLEX_LINCR1	(0x0)
+#define S32G_LINFLEX_LINSR	(0x8)
+#define S32G_LINFLEX_UARTCR	(0x10)
+#define S32G_LINFLEX_UARTSR	(0x14)
+#define S32G_LINFLEX_LINIBRR	(0x28)
+#define S32G_LINFLEX_LINFBRR	(0x24)
+#define S32G_LINFLEX_BDRL	(0x38)
+#define S32G_LINFLEX_UARTPTO	(0x50)
 
 #ifndef __ASSEMBLY__
 struct console_s32g {
diff --git a/plat/s32g/s32g_linflexuart.S b/plat/s32g/s32g_linflexuart.S
index f38ebec86..646f572cf 100644
--- a/plat/s32g/s32g_linflexuart.S
+++ b/plat/s32g/s32g_linflexuart.S
@@ -62,10 +62,10 @@ wait_init_mode:
 
 	/* hard-coded linflexd_serial_setbrg() result,
 	 * working with UARTCR[ROSE]=0
-	 * FIXME: real implementation needed */
-	movz	w16, #72		/* ibr */
+	 */
+	movz	w16, #67		/* ibr */
 	str	w16, [x0, #S32G_LINFLEX_LINIBRR]
-	movz	w16, #5		/* fbr */
+	movz	w16, #13		/* fbr */
 	str	w16, [x0, #S32G_LINFLEX_LINFBRR]
 
 	/* Set preset timeout register value. */
@@ -100,7 +100,7 @@ wait_init_mode:
 	/* prepare to finish console registration */
 	mov	x0, x3
 	mov	x30, x7
-	finish_console_register s32g
+	finish_console_register s32g putc=1, getc=0, flush=1
 
 	movz	w0, 1
 	ret
-- 
2.17.1

